[flang][MIF] Adding Stop and ErrorStop operations

JDPailleux opened the pull request:
This PR proposes to add `Stop` and `ErrorStop` operations to the MIF dialect in order to lower to the corresponding PRIF procedures. 
If the `-fcoarray` flag is passed, then all calls to `STOP` and `ERROR STOP` will be replaced by those of PRIF/MIF. 

Similarly, as specified by the PRIF specification, the program termination must call the PRIF `STOP` procedure to clean up the parallel environment.

JDPailleux committed:
[flang][MIF] Adding Stop and ErrorStop operations

jeanPerier reviewed (changes requested):
Thanks @JDPailleux for working on this topic.
I think a different approach is needed here because:
- program termination may occur in more contexts that STOP/STOP ERROR (e.g. triggered inside the Fortran runtime after invalid IO without IOSTAT/ERRMSG).
- ad-hoc handling of STOP/STOP ERROR in lowering when `-fcoarray` is provided is adding the divergence in the compiler flow earlier than needed, and assumes that all libraries linked to a program using mutli image statements were compiled with `-fcoarray` (I think we should aim to avoid this requirement, although I have done a good job at assessing how much that is possible).

Instead, the Fortran runtime already [has three APIs meant](https://github.com/llvm/llvm-project/blob/d02a5ae10bab57be80bd460d10e6bb30959bdacc/flang-rt/include/flang-rt/runtime/terminator.h#L116) for mutli image Fortran that are called when a program hits terminations:

```
RT_API_ATTRS void NotifyOtherImagesOfNormalEnd();
RT_API_ATTRS void NotifyOtherImagesOfFailImageStatement();
RT_API_ATTRS void NotifyOtherImagesOfErrorTermination();
```

They currently are no-ops.

What I think should happen is that when lowering mif.init:
 - PRIF termination APIs that must be called when the current image terminate should be registered to the Fortran runtime so that they can be called in the three APIs above.
 - Fortran runtime termination should also likely be registered to the PRIF runtime so that when another image fails, the PRIF runtime can initiate the terminations of other images (to make sure that any Fortran runtime data structure/state that needs to be cleaned-up before termination is cleaned-up).
 
So I think you need to add an new API to the Fortran runtime so that you can register the PRIF callbacks as needed. Maybe using `std::atexit` is also an option.

That way, there will be a termination interface between PRIF and the Fortran Runtime without having to catch and special case all termination situations inside compiler code. This should also allow aborts in library compiled without `-focarray` to still properly signal the termination to other images.

JDPailleux committed:
Registering termination functions of PRIF runtime into the Flang runtime

bonachea commented on flang-rt/lib/runtime/terminator.cpp:
As previously discussed, some of the places `NotifyOtherImagesOfNormalEnd` is currently being invoked inside `flang-rt` are probably inappropriate. 

For example the registration as a `std::atexit` handler inside `RTNAME(ProgramStart)` in lib/runtime/main.cpp.  This "just works" today because the callback is a no-op. However once this PR converts it to a parallel synchronization (as required by multi-image `STOP` and implemented in `prif_stop`) this seems likely to create deadlocks.

In a related example, the `RTNAME(StopStatement)` function does not appear to invoke `NotifyOtherImagesOfErrorTermination` at all for the `ERROR STOP` case (it instead prints some output and then calls `std::exit`, which will deadlock upon a non-collective `ERROR STOP`, for the reason above).

Do we want to try and clean this up as part of this PR, or a subsequent PR?

bonachea commented on flang/lib/Lower/Runtime.cpp:
```diff
@@ -105,19 +115,27 @@ void Fortran::lower::genStopStatement(
+    else
+      mif::StopOp::create(builder, loc, stopCode, q);
+  } else
+    fir::CallOp::create(builder, loc, callee, operands);
```
Do we actually need this divergence? (also in Main.cpp)

I thought @jeanPerier was arguing against introducing MIF dialect operations `mif::StopOp`/`mif::ErrorStopOp` at all, and instead just performing all the multi-image "work" inside the callbacks invoked by flang-rt's `RTNAME(StopStatement)/RTNAME(StopStatementText)`.

bonachea reviewed (commented):
@JDPailleux Thanks for the additional work on this! 
This seems like great progress towards the design requested by @jeanPerier .

I've added some questions and noted a likely defect.

klausler commented:
> Thanks for the update, I will have a look. Adding @klausler for the runtime part.

The code in the runtime for image notification is a place-holder; feel free to replace or improve it.  I thought it would be needed by the runtime library's general error termination mechanism (`Terminator`) so that runtime-detected program errors can fail an image gracefully.  My thinking was that since graceful image termination needs to exist for `Terminator`, it could be used as well for `STOP`/`ERROR STOP`/`FAIL IMAGE` and a main program `END` as well.  Having a PRIF implementation call the runtime instead of the other way around might make it harder to use PRIF with non-flang-new Fortran runtimes, if that's a goal.

JDPailleux committed:
Add a structure to manage Exits (ExitHandler) and update the Flang terminaison API

- Configure of the exitHandle base on -fcoarray
- If this option is present, then we use the Flang terminaison API that
  wraps calls to PRIF (prif_stop, prif_error_stop)
- Otherwise, we use the classic std::exit

bonachea commented on flang/lib/Lower/Runtime.cpp:
```diff
@@ -59,6 +60,10 @@ void Fortran::lower::genStopStatement(
   mlir::Location loc = converter.getCurrentLocation();
+  bool coarrayIsEnabled =
+      converter.getFoldingContext().languageFeatures().IsEnabled(
+          Fortran::common::LanguageFeature::Coarray);
```
This variable appears to be unreferenced as of the latest commit.

bonachea commented on flang/lib/Lower/Runtime.cpp:
I don't understand something about the changes made to `Fortran::lower::genStopStatement()` in this file and flang/lib/Optimizer/Builder/Runtime/Main.cpp in the most recent commit [ce614f0](https://github.com/llvm/llvm-project/pull/166787/commits/ce614f0b1d00557cc35077b4c6985a2ad5e181f2).

That commit appears to have removed all invocations of `mif::ErrorStopOp::create()` and `mif::StopOp::create()`, and these don't appear to have been introduced elsewhere. 

Does this mean that those two MIF operations are now effectively "dead" and should also be removed from the MIF dialect? If not, what purpose do they serve and where are these MIF ops being created?

bonachea reviewed (commented):
@JDPailleux Thanks for the updates! They have resolved several of my concerns.

Unfortunately, I have a few new/unresolved concerns with the current state of the PR. I'd be happy to discuss any of these further over Slack or Zoom if that seems helpful.

JDPailleux committed:
Remove useless mif::StopOp and mif::ErrorStopOp + Adding the choose between normal and error exit

bonachea commented on flang-rt/lib/runtime/terminator.cpp:
```diff
@@ -93,10 +100,38 @@ RT_API_ATTRS void Terminator::CrashHeader() const {
+    SynchronizeImagesOfNormalEnd(exitCode);
+  else
+    std::exit(exitCode);
+}
```
The callback invocations are backwards. I assume you instead meant something like the following.

Also, as previously mentioned the PRIF callbacks registered into `normalEndCallback` and `errorCallback` will never return. Unless we want to require/guarantee this of any callbacks registered here, we should perhaps program this more defensively, eg:

```suggestion
void ExitHandler::NormalExit(int exitCode) {
  if (multiImageFeatureEnabled)
    SynchronizeImagesOfNormalEnd(exitCode); // might never return

  std::exit(exitCode);
}

void ExitHandler::ErrorExit(int exitCode) {
  if (multiImageFeatureEnabled)
    NotifyOtherImagesOfErrorTermination(exitCode); // might never return

  std::exit(exitCode);
}
```

bonachea commented on flang/lib/Lower/Runtime.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "flang/Lower/OpenACC.h"
 #include "flang/Lower/OpenMP.h"
 #include "flang/Lower/StatementContext.h"
+#include "flang/Optimizer/Builder/Character.h"
```
This body of this file now contains no changes other than insertion of blank lines, so this include line should probably be removed (and possibly also revert the insertion of blank lines).

bonachea commented on flang-rt/lib/runtime/terminator.cpp:
```diff
@@ -93,10 +100,34 @@ RT_API_ATTRS void Terminator::CrashHeader() const {
+    if (exitCode == EXIT_SUCCESS)
+      SynchronizeImagesOfNormalEnd(exitCode);
+    else
+      NotifyOtherImagesOfErrorTermination(exitCode);
```
Thanks, I think this thread can be resolved now and remaining issues discussed in new threads.

bonachea commented on flang-rt/include/flang-rt/runtime/terminator.h:
```diff
@@ -112,9 +113,23 @@ class Terminator {
+
+  void Configure(bool multiImageFeatureEnabled);
+  void NormalExit(int exitCode);
+  void ErrorExit(int exitCode);
```
Assuming you accept my related suggestion

```suggestion
  [[noreturn]] void NormalExit(int exitCode);
  [[noreturn]] void ErrorExit(int exitCode);
```

bonachea reviewed (commented):
@JDPailleux thanks for the additional improvements!

I've added comments regarding a few additional issues I noticed.

JDPailleux committed:
Update flang-rt/lib/runtime/stop.cpp

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

JDPailleux committed:
Update flang-rt/include/flang-rt/runtime/terminator.h

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

JDPailleux committed:
Update flang-rt/lib/runtime/terminator.cpp

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

JDPailleux commented on flang/lib/Optimizer/Transforms/MIFOpConversion.cpp:
```diff
@@ -74,6 +75,121 @@ static mlir::Value genStatPRIF(fir::FirOpBuilder &builder, mlir::Location loc,
+    }
+    stopCodeInt =
+        fir::AbsentOp::create(builder, loc, builder.getRefType(i32Ty));
+  }
```
Actually, I hadn't noticed that part about QUIET, which isn't really treated anymore since the value is hardcoded. I have no problem to reduce that part and even setting the value to TRUE by default, and removing all treatment for that argument due to the management of QUIET int the flang-rt

JDPailleux committed:
Remove useless header + update ProgramStart call in CommandTest

JDPailleux committed:
Remove QUIET argument in genPRIFStopErrorStop an set quiet to true by default

JDPailleux commented on flang-rt/lib/runtime/stop.cpp:
```diff
@@ -174,17 +174,29 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
 [[noreturn]] void RTNAME(FailImageStatement)() {
   Fortran::runtime::NotifyOtherImagesOfFailImageStatement();
   CloseAllExternalUnits("FAIL IMAGE statement");
```
I have no opinion on whether or not to clear the buffers. However, if reversing the two lines is the desired behavior in multi-image execution, I am not opposed to doing so.

JDPailleux committed:
Update flang-rt/lib/runtime/stop.cpp

Co-authored-by: Dan Bonachea <dobonachea@lbl.gov>

bonachea reviewed (commented):
LGTM!

ktras reviewed (approved):
I have a minor nitpick suggestion listed below. I verified that this code builds and when linked with a PRIF library, supports the Stop and ErrorStop features. Please ensure that you have addressed the earlier comments from @jeanPerier.

jeanPerier commented on flang-rt/lib/runtime/main.cpp:
```diff
@@ -28,8 +28,9 @@ static void ConfigureFloatingPoint() {
-  std::atexit(Fortran::runtime::NotifyOtherImagesOfNormalEnd);
+    const EnvironmentDefaultList *envDefaults,
+    const bool multiImageInitialized) {
+  Fortran::runtime::exitHandler.Configure(multiImageInitialized);
```
Why is this here an not under mif.init lowering?
I think it would be simpler from an engineering point of view if everything is in one place as much as possible.

If the issue is that it is not easy to generate `Fortran::runtime::exitHandler.Configure` call from the compiler, you could have a wrapper API `EnableMultimage` that calls this.

Maybe this is not needed though if you rely on the callback being set-up as the condition.

jeanPerier reviewed (commented):
Thanks, this solves my previous blocking comments and overall looks great!

I have a few comments inlined, mainly requesting to avoid making the globals accessible in a header file which makes them harder to control.

JDPailleux committed:
Remove non necessary boolean in ExitHandler and hide global variables

jeanPerier reviewed (commented):
Two small nits, looks good otherwise.

JDPailleux committed:
Remove exitHandler

jeanPerier reviewed (approved):
Thanks @JDPailleux for addressing all my comments!
